### PR TITLE
Add input prefix and suffix

### PR DESCRIPTION
**Pull request by simonwhatley**
## What

Adds prefix and suffix elements to the input component.

## Why

A number of local and departmental design systems ([GOV.UK Publishing Components](https://components.publishing.service.gov.uk/component-guide/input/with_prefix_and_suffix), [HMRC](https://design.tax.service.gov.uk/hmrc-design-patterns/currency-input/), [MOJ](https://moj-design-system.herokuapp.com/components/currency-input) and [ONS](https://ons-design-system.netlify.app/patterns/numeric-values/)) are using prefixes and suffixes to indicate currencies and measurements.

#### HTML

```html
<div class="govuk-form-group">
  <label class="govuk-label" for="input-with-prefix-attribute">
    Amount
  </label>
  <div class="govuk-input__wrapper">
    <span class="govuk-input__prefix" aria-hidden="true">£</span>
    <input class="govuk-input" id="input-with-prefix-attribute" name="amount" type="number">
  </div>
</div>
```

#### Macro

```
{% from "input/macro.njk" import govukInput %}

{{ govukInput({
  "label": {
    "text": "Amount"
  },
  "id": "input-with-prefix-attribute",
  "name": "amount",
  "type": "number",
  "prefix": {
    "text": "£"
  }
}) }}
```

image

#### HTML

```html
<div class="govuk-form-group">
  <label class="govuk-label" for="input-with-suffix-attribute">
    Weight
  </label>
  <div class="govuk-input__wrapper">
    <input class="govuk-input" id="input-with-suffix-attribute" name="weight" type="number">
    <span class="govuk-input__suffix" aria-hidden="true">kg</span>
  </div>
</div>
```

#### Macro

```
{% from "input/macro.njk" import govukInput %}

{{ govukInput({
  "label": {
    "text": "Weight"
  },
  "id": "input-with-suffix-attribute",
  "name": "weight",
  "type": "number",
  "suffix": {
    "text": "kg"
  }
}) }}
```

image

Fixes https://github.com/alphagov/govuk-frontend/issues/1915

### CODE REVIEW

**Comment by hannalaakso** [comment]
This work was reviewed by the Working Group in June, see https://github.com/alphagov/govuk-design-system/pull/1269#issuecomment-654282576 for their comments.

**Comment by hannalaakso** [comment]
Implements the [design for very narrow viewports](https://github.com/alphagov/govuk-frontend/issues/1915#issuecomment-677705663). It only kicks in at 320px so I think that it's most likely to be seen by users who zoom in on a mobile device. I've tested this using the 'Aa' zoom button on iOS Safari: 

Screenshot 2020-09-01 at 10 45 22
Screenshot 2020-09-01 at 10 45 59
Screenshot 2020-09-01 at 10 46 41
Screenshot 2020-09-01 at 11 09 25
Screenshot 2020-09-01 at 10 47 44
Screenshot 2020-09-01 at 11 35 27

I've also tried to increase text scaling on Android Chrome but it mainly seems to work for text like paragraphs and headings, it doesn't do anything for form elements for instance. I'll raise a separate issue to investigate that.

Tested on 
✅ IE8 (See [commit](https://github.com/alphagov/govuk-frontend/pull/1816/commits/6f8a451b395b5aa9cdbd449eb4edb4b099688335) for details)
❓ IE9 (Doesn't support flex so some borders are incorrectly removed, however number of IE9 users is abysmally small so I'm not worried about it) 
✅ IE10
✅ IE11
✅ Windows 10 Edge 84
✅ Android 10 - Samsung galaxy S20: FF, Chrome, Samsung Internet
✅ Android 6 - Galaxy S7: FF, Chrome
✅ iOS 13 - iPhone XS: Safari
✅ iOS 11 - iPhone 6S: Safari and FF
✅ iOS 9 - iPhone 6S: Safari and FF
❓ iOS 8 - iPhone 6S: Safari and FF (Things aren't quite correctly aligned, probably a flex issue)
❌ iOS 7 - iPhone 5: Safari (Things aren't quite correctly aligned and the suffix drops onto next line - but this is way outside our support matrix and the component is still usable) 
✅ MacOS Safari 13.1
✅ MacOS Firefox 80.0 with inverted colours
✅ MacOS Chrome 84
✅ Windows 10 Firefox 80.0
✅ Windows 10 Edge 84

File: src/govuk/components/input/_index.scss [hunk 1]
```diff
@@ -91,4 +91,88 @@
...
+  .govuk-input__wrapper {
+    display: flex;
+
+    .govuk-input {
+      flex: 0 1 auto;
+    }
+
+    .govuk-input:focus {
+      // Hack to stop focus style being overlapped by the suffix
+      z-index: 1;
```
Comment (hannalaakso) [inline src/govuk/components/input/_index.scss:103]: I'm not keen on this fix but I also can't find another way of stopping the suffix overlapping the input focus border.

File: src/govuk/components/input/input.yaml [hunk 2]
```diff
@@ -54,15 +54,15 @@ params:
     - name: classes
       type: string
       required: false
-      description: Classes to add to the prefix tag
+      description: Classes to add to the prefix.
     - name: attributes
       type: object
       required: false
-      description: HTML attributes (for example data attributes) to add to the prefix tag.
+      description: HTML attributes (for example data attributes) to add to the prefix elementf.
```
Comment (vanitabarrett) [inline src/govuk/components/input/input.yaml:61]:
```suggestion
      description: HTML attributes (for example data attributes) to add to the prefix element.
```

File: src/govuk/components/input/template.test.js [hunk 3]
```diff
@@ -382,4 +389,107 @@ describe('Input', () => {
...
     })
   })
+
+  describe('when it includes a prefix', () => {
+    it('renders the input wrapper', () => {
+      const $ = render('input', examples['with prefix'])
+
+      const $wrapper = $('.govuk-form-group > .govuk-input__wrapper')
+      expect($wrapper.length).toBeTruthy()
+    })
```
Comment (vanitabarrett) [inline src/govuk/components/input/template.test.js:399]: Minor, but the pattern in this file (and I've seen in other test files) is to leave a new line between each test

File: src/govuk/components/input/template.test.js [hunk 4]
```diff
@@ -382,4 +389,107 @@ describe('Input', () => {
...
+    it('renders the prefix inside the wrapper', () => {
+      const $ = render('input', examples['with prefix'])
+
+      const $prefix = $('.govuk-form-group > .govuk-input__wrapper > .govuk-input__prefix')
+      expect($prefix.length).toBeTruthy()
```
Comment (vanitabarrett) [inline src/govuk/components/input/template.test.js:404]: Might be worth checking the prefix text is what we expect too

File: src/govuk/components/input/template.test.js [hunk 5]
```diff
@@ -382,4 +389,107 @@ describe('Input', () => {
...
+      const $ = render('input', examples['with suffix'])
+
+      const $wrapper = $('.govuk-form-group > .govuk-input__wrapper')
+      expect($wrapper.length).toBeTruthy()
+    })
+    it('renders the suffix inside the wrapper', () => {
+      const $ = render('input', examples['with suffix'])
+
+      const $suffix = $('.govuk-form-group > .govuk-input__wrapper > .govuk-input__suffix')
+      expect($suffix.length).toBeTruthy()
```
Comment (vanitabarrett) [inline src/govuk/components/input/template.test.js:451]: Same comment here about checking the suffix text is as expected

**Review by vanitabarrett — CHANGES_REQUESTED** [review]

File: src/govuk/components/input/template.njk [hunk 6]
```diff
@@ -5,6 +5,18 @@
...
+
+{%- set inputTag %}
+<input class="govuk-input {%- if params.classes %} {{ params.classes }}{% endif %} {%- if params.errorMessage %} govuk-input--error{% endif %}" id="{{ params.id }}" name="{{ params.name }}" type="{{ params.type | default('text') }}"
+{%- if (params.spellcheck === false) or (params.spellcheck === true) %} spellcheck="{{ params.spellcheck }}"{% endif %}
+{%- if params.value %} value="{{ params.value}}"{% endif %}
+{%- if describedBy %} aria-describedby="{{ describedBy }}"{% endif %}
+{%- if params.autocomplete %} autocomplete="{{ params.autocomplete}}"{% endif %}
+{%- if params.pattern %} pattern="{{ params.pattern }}"{% endif %}
+{%- if params.inputmode %} inputmode="{{ params.inputmode }}"{% endif %}
+{%- for attribute, value in params.attributes %} {{ attribute }}="{{ value }}"{% endfor -%}>
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply src/govuk/components/input/template.njk]: This block isn't modified in this PR (I think you were looking at part of commit [`be2c2de` (#1816)](https://github.com/alphagov/govuk-frontend/pull/1816/commits/be2c2debc2d3fec8e7a3b11c8b242ff1d44be18c) which is outdated).

File: src/govuk/components/input/_index.scss [hunk 7]
```diff
@@ -91,4 +91,56 @@
...
+    }
+  }
+
+  .govuk-input__prefix,
+  .govuk-input__suffix {
+    @include govuk-font($size: 19);
+
+    box-sizing: border-box;
+    display: inline-block;
+    width: auto;
```
[the remark this answers is not in the document]
Comment (hannalaakso) [reply src/govuk/components/input/_index.scss]: Good spot 👍

File: src/govuk/components/input/template.test.js (hunk 4 above)
Comment (hannalaakso) [reply src/govuk/components/input/template.test.js:405]: In my I had covered this with `allows prefix text to be passed whilst escaping HTML entities` but I've added a new test to check the text in the default example, it's more explicit 👍

File: src/govuk/components/input/template.test.js (hunk 5 above)
Comment (hannalaakso) [reply src/govuk/components/input/template.test.js:466]: Same as above but with `allows suffix text to be passed whilst escaping HTML entities`.

**Review by vanitabarrett — APPROVED** [review]

**Comment by 36degrees** [comment]
We want to hold off merging this until we're ready to start on 4.0.

**Comment by hannalaakso** [comment]
I'll also need to draft something for the changelog, will do that shortly

**Comment by 36degrees** [comment]
Minor, but can we have at least one example that does not use an input width modifier? This will allow us to verify that the input is unconstrained and 'full-width' by default.

Possibly most of the examples should omit the width modifier? In the same way the National Insurance Number examples are 'full-width' even though in real world usage you'd expect them to be constrained… most of the examples should probably combine the minimum number of 'features' possible? Although we would then want some examples that show that they do combine properly…

**Comment by hannalaakso** [comment]
@36degrees I don't know if you saw the [feedback](https://github.com/alphagov/govuk-design-system/pull/1269#issuecomment-684966546) from @christopherthomasdesign which the change with the width modifiers was in response to? 

But maybe we mainly shouldn't use the width modifiers in the review app as the examples are out of context anyway..? Let me know if you have any thought on that. Definitely happy to include some examples without width modifiers in either case.

**Comment by 36degrees** [comment]
> @36degrees I don't know if you saw the [feedback](https://github.com/alphagov/govuk-design-system/pull/1269#issuecomment-684966546) from @christopherthomasdesign which the change with the width modifiers was in response to?

Yep, I did – and completely agree for the examples in the Design System.

> But maybe we mainly shouldn't use the width modifiers in the review app as the examples are out of context anyway..? Let me know if you have any thought on that. Definitely happy to include some examples without width modifiers in either case.

I think _generally_ we use examples that are focused on only the specific feature, and don't 'enable' anything else – for example, the National Insurance Number examples don't include the spellcheck attribute, the example that enables isPageHeading doesn't also add a class to style it (though I think we might be missing some examples for each label size!)

**Comment by hannalaakso** [comment]
That all makes sense 👍 

However if you remove the width modifiers altogether you end up with this that to my eye looks pretty strange (I hadn't realised this up until point either):

Screenshot 2020-09-03 at 18 20 17

Or is it just me and they actually look okay like that? I've been staring at them for so long over the past weeks that I can't tell anymore 😆

**Comment by edwardhorsford** [comment]
> That all makes sense 👍
> 
> However if you remove the width modifiers altogether you end up with this that to my eye looks pretty strange (I hadn't realised this up until point either):
> 
> Screenshot 2020-09-03 at 18 20 17
> 
> Or is it just me and they actually look okay like that? I've been staring at them for so long over the past weeks that I can't tell anymore 😆

Do these end up showing in a 2/3 column grid?

**Comment by hannalaakso** [comment]
> Do these end up showing in a 2/3 column grid?

They get constrained by the grid so like I said I think we're okay.

**Comment by hannalaakso** [comment]
Massive thanks for all the amazing work by @simonwhatley and his team on this component 🎉🙏

**Pull request merged**